File: 0_1.baekjoon/dp/java/src/BOJ21941.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.StringTokenizer;
public class BOJ21941 {
    private static String S;
    private static int N;
    private static Map<String, Integer> words;
    private static int[] dp;
    public static void main(String args[]) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        words = new HashMap<>();
        S = br.readLine();
        N = S.length();
        dp = new int[N];
        int M = Integer.parseInt(br.readLine());

        Arrays.fill(dp, -1);
        for (int i=0;i<M;i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            String word = st.nextToken();
            int point = Integer.parseInt(st.nextToken());

            if (word.length() >= point) continue;
            words.put(word, Math.max(words.getOrDefault(word, 0), point));
        }

        System.out.println(dfs(0));

    }

    private static int dfs(int idx) {
        if (idx >= N) {
            return 0;
        }

        if (dp[idx] != -1) {
            return dp[idx];
        }

        int result = 0;

        for (String word : words.keySet()) {
            int length = word.length();
            if (S.startsWith(word, idx)) {
                dp[idx] = Math.max(dp[idx], dfs(idx + length) + words.get(word));
            }
        }

        dp[idx] = Math.max(dp[idx], dfs(idx + 1) + 1);

        return dp[idx];
    }
}

File: 0_1.baekjoon/dp/kotlin/src/BOJ21941.kt
import kotlin.math.max

private lateinit var dp: IntArray
private lateinit var wordMap: MutableMap<String, Int>
private var s = ""
fun main() = with(System.`in`.bufferedReader()) {
    s = readLine()
    wordMap = mutableMapOf<String, Int>()
    repeat(readLine().toInt()) {
        val (word, value) = readLine().split(" ")
        wordMap[word] = value.toInt()
    }
    dp = IntArray(s.length + 1)
    print(dfs(s.length, 0))
}

private fun dfs(limit: Int, idx: Int): Int {
    if (idx >= limit) {
        return 0
    }

    if (dp[idx] != 0) {
        return dp[idx]
    }

    wordMap.keys.forEach {
        val size = it.length
        if (s.startsWith(it, idx)) {
            dp[idx] = max(dp[idx], dfs(limit, idx + size) + wordMap[it]!!)
        }
    }
    dp[idx] = max(dp[idx], dfs(limit, idx + 1) + 1)

    return dp[idx]
}